module/zfs/spa.c

```diff
@@ -1669,23 +1707,95 @@ spa_thread(void *arg)
 }
 #endif
 
-extern metaslab_ops_t *metaslab_allocator(spa_t *spa);
+/*
+ * Returns with the spa_chain_map_lock held. This prevents the shared log
+ * pool from being exported or deleted while a pool is being activated that
+ * depends on it.
+ */
+static int
+get_shared_log_pool(nvlist_t *config, spa_t **out)
+{
+	if (config == 0)
+		return (0);
+	uint64_t guid;
+	if (nvlist_lookup_uint64(config, ZPOOL_CONFIG_SHARED_LOG_POOL, &guid))
+		return (0);
+	spa_t *search = kmem_zalloc(sizeof (spa_t), KM_SLEEP);
+	search->spa_config_guid = guid;
+	mutex_enter(&spa_shared_log_lock);
+	spa_t *result = avl_find(&spa_shared_log_avl, search, NULL);
+	kmem_free(search, sizeof (*search));
+	if (!result) {
+		mutex_exit(&spa_shared_log_lock);
+		return (ENOENT);
+	}
+	mutex_enter(&result->spa_chain_map_lock);
+	mutex_exit(&spa_shared_log_lock);
+	*out = result;
+
+	avl_tree_t *t = &result->spa_chain_map;
+	spa_chain_map_pool_t *search_scmp = kmem_zalloc(sizeof (*search_scmp),
+	    KM_SLEEP);
+	if (nvlist_lookup_uint64(config, ZPOOL_CONFIG_POOL_GUID, &guid))
+		return (0);
```

Comment:
Can pool really have no GUID?

Reply:
Weirdly, yes! I don't remember if it's a specific path in tryimport or some other odd case, but it can happen.